src/v/security/role_store.h

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/container/node_hash_map.h"
+#include "security/role.h"
+#include "security/types.h"
+
+#include <seastar/util/variant_utils.hh>
+
+#include <ranges>
+#include <type_traits>
+
+namespace security {
+
+/* container for Roles */
+class role_store {
+    using role_types = std::variant<role>;
+    using container_type = absl::node_hash_map<role_name, role_types>;
+    using member_cache_type
+      = absl::node_hash_map<role_member, absl::flat_hash_set<role_name>>;
+
+public:
+    role_store() noexcept = default;
+    role_store(const role_store&) = delete;
+    role_store& operator=(const role_store&) = delete;
+    role_store(role_store&&) noexcept = default;
+    role_store& operator=(role_store&&) noexcept = default;
+    ~role_store() noexcept = default;
+
+    template<typename T>
+    bool put(const role_name& name, T&& role) {
+        auto res = _roles.insert_or_assign(name, std::forward<T>(role));
+        using v_type = typename std::remove_cv<
+          typename std::remove_reference<T>::type>::type;
+        for (const auto& m : std::get<v_type>(res.first->second).members()) {
+            _member_cache[m].insert(name);
+        }
+        return res.second;
+    }
+
+    template<typename T>
+    std::optional<T> get(const role_name& name) const {
+        if (auto it = _roles.find(name); it != _roles.end()) {
+            return std::get<T>(it->second);
+        }
+        return std::nullopt;
+    }
+
+    std::optional<const member_cache_type::value_type::second_type*>
+    get_member_roles(const role_member& user) const {
+        auto it = _member_cache.find(user);
+        if (it != _member_cache.end()) {
+            return &it->second;
+        }
+        return std::nullopt;
+    }
```

Comment:
Not particularly wedded to it, but I'm hesitant to foreclose on the `role_name -> set<member>` model for some reason. Maybe just because that's what it _is_ in an ontological sense. 

Open to any suggestions to the contrary, esp. anything that might reduce the memory footprint of this structure.

Reply:
Perhaps something like this is nicer (I have fixups if you want them):
```suggestion
    boost::iterator_range<
      member_cache_type::value_type::second_type::const_iterator>
    get_member_roles(const role_member& user) const {
        auto it = _member_cache.find(user);
        if (it != _member_cache.end()) {
            return it->second;
        }
        return {};
    }
```
I don't think it's worth maintaining two containers here; I usually reach for boost multi_index_container when I have this type of requirement, but it's not really suitable here (and won't be until the hashed indexes are updated with the new open addressing container impl from boost container).

Reply:
>`boost::iterator_range`

Ah, yeah, I spent like an hour last night wracking my brain thinking there was something like that in `std`. Great!

>don't think it's worth maintaining two containers here

Neither do I. Probably fine to have the put/get interface emit `role` objects that are constructed on demand from the state of the members collection, no? Do you have something more clever?

Reply:
Incidentally, I don't think multi_index is appropriate at all, right? The issue is that we have two different aggregates that we might want to operate on: a list of role names OR a set of role members. The former for authZ and the latter for "management". I might be over-fixated on the latter abstraction though...

Reply:
> Probably fine to have the put/get interface emit role objects that are constructed on demand from the state of the members collection

Yeah, they're value types, at least!

Reply:
true indeed 👍